compiler/next/include/chpl/Queries/ContextDetail.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_QUERIES_CONTEXT_DETAIL_H
+#define CHPL_QUERIES_CONTEXT_DETAIL_H
+
+#include "chpl/AST/UniqueString.h"
+#include "chpl/AST/ErrorMessage.h"
+#include "chpl/Util/memory.h"
+
+#include <cstring>
+#include <unordered_map>
+#include <vector>
+
+/// \cond DO_NOT_DOCUMENT
+namespace chpl {
+namespace detail {
+
+struct UniqueStrEqual final {
+  bool operator()(const char* lhs, const char* rhs) const {
+    // pass the 2 metadata bytes
+    return strcmp(lhs, rhs) == 0;
+  }
+};
+
+struct UniqueStrHash final {
+  std::size_t operator()(const char* s) const {
+    // this hash is from StringHashFns in the old map.h
+    unsigned int h = 0;
+    while (*s) h = h * 27 + (unsigned char)*s++;
+    return h;
+  }
+};
+
+}
+
+namespace querydetail {
+
+typedef int64_t RevisionNumber;
+class QueryMapResultBase;
+
+// define a hash function for std::tuple since the standard doesn't
+// include one.
+template<typename T>
+static std::size_t queryArgsHashOne(const T& v) {
+  std::hash<T> hasher;
+  return hasher(v);
+}
+
```

Comment:
Look at https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/library/src/include/tuple_helper.hpp for how I implemented tuple hashes (it was in C++14 though, such as using `std::index_sequence` to avoid recursion).